src/client/dfuse/pil4dfs/dfs_dcache.c

```diff
@@ -0,0 +1,616 @@
+#include <fcntl.h>
+
+#include <daos/debug.h>
+#include <gurt/hash.h>
+#include <gurt/list.h>
+#include <daos_fs.h>
+#include <daos/common.h>
+
+#include "dfs_dcache.h"
+
+/** Size of the hash key prefix */
+#define DCACHE_KEY_PREF_SIZE 35
+
+typedef int (*destroy_fn_t)(dfs_dcache_t *);
+typedef int (*find_insert_fn_t)(dcache_rec_t **, dfs_dcache_t *, char *, size_t);
+typedef void (*drec_incref_fn_t)(dfs_dcache_t *, dcache_rec_t *);
+typedef void (*drec_decref_fn_t)(dfs_dcache_t *, dcache_rec_t *);
+typedef void (*drec_del_at_fn_t)(dfs_dcache_t *, dcache_rec_t *);
+typedef int (*drec_del_fn_t)(dfs_dcache_t *, char *, dcache_rec_t *);
+
+/** DFS directory cache */
+struct dfs_dcache {
+	/** Cached DAOS file system */
+	dfs_t              *dd_dfs;
+	/** Hash table holding the cached directories */
+	struct d_hash_table dd_dir_hash;
+	/** Key prefix of the DFS root directory */
+	char                dd_key_root_prefix[DCACHE_KEY_PREF_SIZE];
+	/** Destroy a dfs dir-cache */
+	destroy_fn_t        destroy_fn;
+	/** Return the dir-cahe record of a given location and insert it if needed */
+	find_insert_fn_t    find_insert_fn;
+	/** Increase the reference counter of a given dir-cache record */
+	drec_incref_fn_t    drec_incref_fn;
+	/** Decrease the reference counter of a given dir-cache record */
+	drec_decref_fn_t    drec_decref_fn;
+	/** Delete a given dir-cache record */
+	drec_del_at_fn_t    drec_del_at_fn;
+	/** Delete the dir-cache record of a given location */
+	drec_del_fn_t       drec_del_fn;
+};
+
+/** Entry of a DFS directory cache */
+struct dcache_rec {
+	/** Entry in the hash table of the DFS cache */
+	d_list_t         dr_entry;
+	/** Cached DFS directory */
+	dfs_obj_t       *dr_obj;
+	/** Reference counter used to manage memory deallocation */
+	_Atomic uint32_t dr_ref;
+	/** True iff this entry was deleted */
+	_Atomic bool     dr_deleted;
+	/** Key prefix used by its child directory */
+	char             dr_key_child_prefix[DCACHE_KEY_PREF_SIZE];
+	/** Length of the hash key used to compute the hash index */
+	size_t           dr_key_len;
+	/** the hash key used to compute the hash index */
+	char             dr_key[];
+};
+
+static inline dcache_rec_t *
+dlist2drec(d_list_t *rlink)
+{
+	return d_list_entry(rlink, dcache_rec_t, dr_entry);
+}
+
+static bool
+dcache_key_cmp(struct d_hash_table *htable, d_list_t *rlink, const void *key, unsigned int key_len)
+{
+	dcache_rec_t *rec;
+
+	rec = dlist2drec(rlink);
+	if (rec->dr_key_len != key_len)
+		return false;
+	D_ASSERT(rec->dr_key[key_len] == '\0');
+	D_ASSERT(((char *)key)[key_len] == '\0');
+
+	return strncmp(rec->dr_key, (const char *)key, key_len) == 0;
+}
+
+static void
+dcache_rec_free(struct d_hash_table *htable, d_list_t *rlink)
+{
+	int           rc;
+	dcache_rec_t *rec;
+
+	rec = dlist2drec(rlink);
+	D_DEBUG(DB_TRACE, "delete record '%s' (ref=%u)", rec->dr_key, atomic_load(&rec->dr_ref));
+	rc = dfs_release(rec->dr_obj);
+	if (rc == ENOMEM)
+		rc = dfs_release(rec->dr_obj);
```

Comment:
I took this code from the original dir-entry cache code, which was looking strange for me.
Thus, I am not sure from where it was taken, maybe @wiliamhuang could have more info on this point.

- [ ] Remove retry dfs_release()

Reply:
Indeed, I previously found a few places in other code retrying dfs_release() to avoid fault injection testing issues more than a half year ago. I copied it into libpil4dfs.
I just went through all code with dfs_release(). It looks like all of them are fixed now except in int_dfs.c.

Reply:
> I took this code from the original dir-entry cache code, which was looking strange for me. Thus, I am not sure from where it was taken, maybe @wiliamhuang could have more info on this point.
> 
> * [x]  Remove retry dfs_release()

Fixed with commit e6dd70d